Implementation of held/unheld functions for state pkg

vzhestkov opened the pull request:
### What does this PR do?

Implements `held`/`unheld` functions for state `pkg` module.
Examples:
```
sysstat:
  pkg.held: []
```
```
held-pkgs:
  pkg.held:
    - pkgs:
      - sysstat
      - vnstat: 1.17-bp153.1.18
```
```
held-pkgs-clear:
  pkg.held:
    - pkgs: []
    - replace: True
```
```
unheld-pkgs:
  pkg.unheld:
    - pkgs:
      - sysstat
      - vnstat: 1.17-bp153.1.18
      - test-package
```
```
unheld-all:
  pkg.unheld:
    - all: True
```

Implements `list_holds` function for `zypperpkg` module. The output is similar to `yumpkg.list_holds`. The only difference is that `zypperpkg` calculates the current locked packages from `list_locks` as the yum and zypper have different way of managing locks.

Improves `hold` function for `zypperpkg` to handle package version if specified.

vzhestkov committed:
Implementation of held/unheld functions for state pkg

Fix docs for held/unheld functions of pkg state

Remove ignoring retcode 104 in favour for PR#58552

Add `root` parameter to hold/unhold functions

Move `root` to functions signature

and add missing parameter description to held/unheld

Remove all_versions and ignore_no_matching_item

Add root to hold/unhold docstrings

Remove empty line

Ch3LL reviewed (changes requested):
this will require a changelog and test coverage

vzhestkov committed:
Add unit tests for hold/unhold/list_holds of zypperpkg

Add unit test for pkg.held/unheld state functions

Code formatting fix for the tests

Fix test formatting

vzhestkov committed:
Add changelog entry

Ch3LL commented on salt/modules/zypperpkg.py:
```diff
@@ -2106,28 +2179,42 @@ def unhold(name=None, pkgs=None, **kwargs):
 
     targets = []
     if pkgs:
-        for pkg in salt.utils.data.repack_dictlist(pkgs):
```
Why are you removing this?

vzhestkov replied:
This for loop removes the versions from the `pkgs` list, but in `zypper` the package could be locked with the specified version

Ch3LL commented on salt/modules/zypperpkg.py:
```diff
@@ -2205,18 +2295,20 @@ def hold(name=None, pkgs=None, **kwargs):
 
     targets = []
     if pkgs:
-        for pkg in salt.utils.data.repack_dictlist(pkgs):
```
here as well. why removing?

vzhestkov replied:
same as above

Ch3LL commented on salt/modules/zypperpkg.py:
```diff
@@ -2558,6 +2650,9 @@ def search(criteria, refresh=False, **kwargs):
     details (bool)
         Show version and repository
 
+    ignore_no_matching_item (bool)
```
I'm not seeing this kwarg being used in this function.

vzhestkov commented on salt/modules/zypperpkg.py:
```diff
@@ -2007,6 +2007,76 @@ def purge(name=None, pkgs=None, root=None, **kwargs):  # pylint: disable=unused-
+                case_sensitive=(lock.get("case_sensitive", "on") == "on"),
+                installed_only=True,
+                details=True,
+                all_versions=True,
```
Just realized that `all_version` left there as `info_installed` was used before, but then I've changed it to `search`.
`ignore_no_matching_item` is a side effect of implementing the fix which is actually should be fixed with this PR:https://github.com/saltstack/salt/pull/58552 I reverted the change required to fix this issue, but not fully. Will fix it. Thanks

vzhestkov replied:
Fixed with https://github.com/saltstack/salt/pull/60432/commits/a685243947cfcd383b60101ab5c4f2afd7f408e6

twangboy commented on salt/modules/zypperpkg.py:
```diff
@@ -2018,6 +2018,73 @@ def purge(name=None, pkgs=None, root=None, **kwargs):  # pylint: disable=unused-
+def list_holds(pattern=None, full=True, root=True, **kwargs):
+    """
+    List information on locked packages.
+
```
This function needs a `.. versionadded:: Phosphorus` line in the docs

twangboy replied:
If you want this to go in for Phosphorus, we need to have versionadded. But it's 3005 now so:
```
.. versionadded:: 3005
```

twangboy commented on salt/states/pkg.py:
```diff
@@ -3629,3 +3629,313 @@ def mod_beacon(name, **kwargs):
+
+def held(name, version=None, pkgs=None, replace=False, **kwargs):
+    """
+    Set package in 'hold' state, meaning it will not be changed.
```
This new function also needs a `.. versionadded:: Phosphorus` docs marker

twangboy replied:
If you want this to go in for Phosphorus, we need to have versionadded. But it's 3005 now so:
```
.. versionadded:: 3005
```

twangboy commented on salt/states/pkg.py:
```diff
@@ -3629,3 +3629,313 @@ def mod_beacon(name, **kwargs):
+
+def unheld(name, version=None, pkgs=None, all=False, **kwargs):
+    """
+    Unset package from 'hold' state, to allow operations with the package.
```
`.. versionadded:: Phosphorus`

twangboy replied:
If you want this to go in for Phosphorus, we need to have versionadded. But it's 3005 now so:
```
.. versionadded:: 3005
```

garethgreenaway commented:
@vzhestkov Some merge conflicts on this one and some requested changes.

vzhestkov committed:
Blacken test_pkg

twangboy committed:
Add versionadded

vzhestkov commented:
@twangboy thanks a lot for taking care of it

garethgreenaway committed:
Merge branch 'master' into held-unheld-implementation
